extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -309,9 +338,69 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
   }
 
   /** @override */
+  buildCallback() {
+    super.buildCallback();
+
+    if (this.isResponsive()) {
+      // Attempt to resize to the correct height.
+      const viewport = this.getViewport();
+      this.responsiveSizeChangePromise_ =
+          this.attemptChangeSize(
+              AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
+                  viewport.getSize()),
+              undefined);
+    }
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    super.onLayoutMeasure();
+
+    if (this.isResponsive() && !this.responsiveAligned_) {
+      this.responsiveAligned_ = true;
+
+      const layoutBox = this.getLayoutBox();
+
+      // Nudge into the correct horizontal position. NB: this must succeed in
+      // order to align the element correctly, so we use changeSize rather than
+      // attemptChangeSize here. But it doesn't actually change the size (it
+      // just changes the horizontal margins).
+      // TODO(charliereams): This is wrong for RTL.
+      this.element.getResources()./*OK*/changeSize(
+          this.element, undefined, undefined, undefined,
+          {left: -1 * layoutBox.left});
+      setStyle(this.element, 'zIndex', 30);
+    }
+  }
+
+  /** @override */
+  layoutCallback() {
+    const callback = super.layoutCallback();
+    if (this.responsiveSizeChangePromise_) {
+      return callback.then(this.responsiveSizeChangePromise_);
+    }
+    return callback;
+  }
+
+  /** @override */
   getPreconnectUrls() {
     return ['https://googleads.g.doubleclick.net'];
   }
+
+  /**
+   * Calculates the appropriate height for a full-width responsive ad of the
+   * given width.
+   * @param {!{width: number, height: number}} viewportSize
+   * @return {number}
+   * @private
+   */
+  static getResponsiveHeightForContext_(viewportSize) {
```

Comment:
Is there a reason this can't be done entirely with a `[heights]` attribute?

```html
<amp-ad heights="(max-height: 120px) 100px, (min-height: 360px) 300px, 83%">
</amp-ad>
```